### ISSUE TITLE
How to set up custom regex for parserOpts ?

### ISSUE THREAD

**Issue by jy95** [semantic-release/semantic-release#575]
Hello everyone,

I use the conventional-changelog preset eslint with [sgc ](https://github.com/JPeer264/node-semantic-git-commit-cli) and I would to handle a custom thing introduced by sgc : emoji at the beginning of the commit title.

Now, a build where I use [that ](https://travis-ci.org/jy95/torrent-files-library/jobs/320472340): 

```
[Semantic release]: Call plugin analyze-commits
[Semantic release]: Analyzing commit: :hammer: Refactor: Update description inside package.json
[Semantic release]: The commit should not trigger a release
``` 

and a build without [that ](https://travis-ci.org/jy95/torrent-files-library/jobs/320475836): 

```
[Semantic release]: Call plugin analyze-commits
[Semantic release]: Analyzing commit: Refactor: Remove emoji for sgcrc
[Semantic release]: The release type for the commit is patch 
``` 

My current config in package.json (and the [rules ](https://github.com/jy95/torrent-files-library/blob/master/config/release-rules.js))
```json
"release": {
    "analyzeCommits": {
      "preset": "eslint",
      "releaseRules": "./config/release-rules.js"
    },
    "generateNotes": {
      "preset": "eslint"
    }
  }
``` 

The settings I would like to add (for analyzeCommits and generateNotes) : 

``` 
"parserOpts": {
        "headerPattern": /^(?:\:(\w*)\:)?\s(\w*)\:\s(.*?)(?:\((.*)\))?$/,
        "headerCorrespondence": [
          "emoji",
          "tag",
          "message"
        ]
      }
``` 

The problem is that [JSON ](http://json.org/) file (package.json) allows only primitive values.

So, I think I could be nice to provide a example of **release.config.js** based on this case, isn't it ? (since the current tests doesn't show one )

Thanks for helping.

**Comment by pvdlg** [comment]
Thanks for the feedback and suggestion. Could you just clarify if what you are trying to achieve works and if you are suggesting to add more examples to the doc, or if you are asking for help to make it work?

More generally, the `parserOpts` in `analyzeCommits` and `generateNotes` is there for convenience to allow you to make minor changes to a conventional-changelog preset.

If you want to achieve something more advanced the best option is to implement your own conventional-changelog preset. Here is an example I did a while ago (It's not really up to date but can be of some help): https://github.com/pvdlg/conventional-changelog-metahub.

If you name your preset module `conventional-changelog-my-preset`, then you can configure `semantic-release` as follow:
```json
"release": {
    "analyzeCommits": {
      "preset": "my-preset",
      "releaseRules": "./config/release-rules.js"
    },
    "generateNotes": {
      "preset": "my-preset"
    }
  }
```

Technically you could do all of that in a `release.config.js` file, but it's probably easier to create a preset. In addition it makes it easier to re-use.

That said, we can add a **Note** in the README of both plugins, linking to the `semantic-release` README part mentioning the different config file type, explaining that it can allow to set `regex` or `function` property if necessary. But in most cases, creating a `preset` is the recommended way.

PR welcomes for the doc additions!

**Comment by jy95** [comment]
To clarify I just want my first build to trigger the release like the second build but the preset cannot handle the little change added by [sgc](https://github.com/JPeer264/node-semantic-git-commit-cli) (refered in the readme)

Recreate a preset that looks like 99% of the [conventional-changelog-eslint](https://github.com/conventional-changelog-archived-repos/conventional-changelog-eslint) package looks like a quick fix but not the best solution (think about updates that fix parts of the code I didn't touch ) ....

I just think strange that **release.config.js** is never used anywhere ^^

**Comment by pvdlg** [comment]
Ah ok, thanks for clarification.

Just a FYI, you can create your own preset by "extending" and existing one, this way you don't have to rewrite everything.

If you want to set a regexp option in a `semantic-release` config you have two options:
- Use `release.config.js` instead of the `release` property in`package.json`
- Create a shareable config (implemented in #553)

`release.config.js` is not mentioned in most docs as it's a feature that has been implemented recently, after most docs were writing.

I'm not really sure where the problem is. Does that works the way you want when you use `release.config.js`?

**Comment by jy95** [comment]
Thanks for the detailed answer.
I think It could be useful to have a example with **release.config.js** since it is new XD

I would like to test it  (using it in local with debug) but I didn't know how to use it ^^

From what I have understand, : 

**release.config.js**
```js
/* eslint-disable no-useless-escape */
module.exports = {
  analyzeCommits: {
    preset: 'eslint',
    releaseRules: './config/release-rules.js',
    parserOpts: {
      headerPattern: /^(?:\:(\w*)\:)?\s(\w*)\:\s(.*?)(?:\((.*)\))?$/,
      headerCorrespondence: [
        'emoji',
        'tag',
        'message',
      ],
    },
  },
  generateNotes: {
    preset: 'eslint',
    parserOpts: {
      headerPattern: /^(?:\:(\w*)\:)?\s(\w*)\:\s(.*?)(?:\((.*)\))?$/,
      headerCorrespondence: [
        'emoji',
        'tag',
        'message',
      ],
    },
  },
};
```

and to set up inside semantic-release, the only option seems to be

```
Usage: semantic-release [options]

  Options:
  -e, --extends <paths>    Comma separated list of shareable config paths or packages name
```

So
```bash
semantic-release -e ./config/release.config.js --debug
```

gives : 

```bash
  semantic-release:get-version-head Check if the current working directory is a git repository +0ms
  semantic-release:get-version-head cmd: 'git rev-parse --git-dir' +1ms
  semantic-release:get-version-head stdout: '.git' +2ms
  semantic-release:get-version-head stderr: '' +0ms
  semantic-release:get-version-head code: 0 +0ms
  semantic-release:config options values: [ 'branch',
  semantic-release:config   'repositoryUrl',
  semantic-release:config   'analyzeCommits',
  semantic-release:config   'generateNotes',
  semantic-release:config   'debug' ] +0ms
  semantic-release:config name: undefined +2ms
  semantic-release:config branch: 'master' +0ms
  semantic-release:config repositoryUrl: 'git+https://github.com/jy95/torrent-files-library-cli.git' +0ms
  semantic-release:config analyzeCommits: { preset: 'eslint',
  semantic-release:config   releaseRules: './config/release-rules.js',
  semantic-release:config   parserOpts:
  semantic-release:config    { headerPattern: /^(?:\:(\w*)\:)?\s(\w*)\:\s(.*?)(?:\((.*)\))?$/,
  semantic-release:config      headerCorrespondence: [ 'emoji', 'tag', 'message' ] } } +0ms
  semantic-release:config generateNotes: { preset: 'eslint',
  semantic-release:config   parserOpts:
  semantic-release:config    { headerPattern: /^(?:\:(\w*)\:)?\s(\w*)\:\s(.*?)(?:\((.*)\))?$/,
  semantic-release:config      headerCorrespondence: [ 'emoji', 'tag', 'message' ] } } +1ms
  semantic-release:config verifyConditions: undefined +0ms
  semantic-release:config verifyRelease: undefined +1ms
  semantic-release:config publish: undefined +0ms
[Semantic release]: Load plugin verifyConditions from @semantic-release/npm
[Semantic release]: Load plugin verifyConditions from @semantic-release/github
[Semantic release]: Load plugin verifyConditions from @semantic-release/condition-travis
[Semantic release]: Load plugin getLastRelease from @semantic-release/npm
[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer
[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator
[Semantic release]: Load plugin publish from @semantic-release/npm
[Semantic release]: Load plugin publish from @semantic-release/github
```

I will test it on my cli project to see if it works (but It looks like good)
